[RFC] cp: major overhaul of the page

waldyrious opened the pull request:
- change "directory" to "folder" (more beginner-friendly)
- change "original" --> "copy" to "file"/"folder" --> "copy", to be more explicit (inspired by #289)
- use an extension to distinguish file paths from folder paths
- reword some of the descriptions for clarity
- include examples to clarify when the target argument defines the actual copy, and when it defines the folder where to put the copy
- add an example to copy contents of a folder into another (existing) folder

waldyrious commented:
I'm marking this as RFC because these changes make the page too long, so we might want to trim down some of the examples.

ostera commented on pages/common/cp.md:
```diff
@@ -1,27 +1,35 @@
+`cp {{/path/to/file.ext}} {{/path/to/copy.ext}}`
 
-- Copy a file to a parent directory:
+- Copy a file into another folder, keeping the filename:
```
This and the first one are redundant IMO.

waldyrious replied:
As I said in the opening comment, I've added both examples "to clarify when the target argument defines the actual copy, and when it defines the folder where to put the copy". Don't you think the second usage may not be clear if we don't explicitly mention it?

agnivade replied:
Gave some thought to this. Have to agree with @ostera on this. IMO, the second usage is easily deducible from the first.

waldyrious replied:
Fair enough, I can see your points -- I'll remove it.

ostera commented on pages/common/cp.md:
```diff
@@ -1,27 +1,35 @@
+
+- Copy the contents of a folder into another folder:
+
+`cp -r {{/path/to/source/folder/*}} {{/path/to/target/folder}}`
```
I think this one can be scrapped, but I have no strong arguments against keeping it. It just feels a bit cumbersome to have 3 recursive-copy examples.

waldyrious replied:
I added because this was what prompted me to make this change :P but I'm ok with removing it, as it doesn't exactly cover a cp-specific functionality, despite its (subjective) usefulness.

ostera replied:
I mean this could be covered by `cp -r path1 path2; mv path2 path3`, but I see what you mean. Could we remove the one that keeps the folder name maybe?

waldyrious replied:
I'd like to keep that one for symmetry with the file-based examples, and for completeness :( I'm honestly conflicted as to what to remove (hence the RFC)

agnivade replied:
3 examples with `-r` does seem a bit too much :smile:. Honestly, I always get stumped over whether to add the trailing / at the end of a folder or not. I would just keep your first and last example(the one with /*) and be done with it.

waldyrious replied:
From my tests, the trailing slash makes no difference by itself -- in fact, adding a trailing slash to the source directory actually seems to cause the copying not to happen (tested on OS X, if it matters).

waldyrious replied:
@agnivade does that mean you'd remove the "Copy a folder recursively into another folder, keeping the folder name" example? I think the file version of this example is deducible, but this one is less intuitive. I may be biased though as I recently had to do it, so your comments are welcome. How about we keep that one and merge the -v with the `/*` example?

agnivade replied:
>  in fact, adding a trailing slash to the source directory actually seems to cause the copying not to happen

Wow .. interesting.

>  How about we keep that one and merge the -v with the /* example?

I'm okay with it. Since we are removing it from the plain copy example, its okay to keep it here.

ostera commented on pages/common/cp.md:
```diff
@@ -1,27 +1,35 @@
+
+`cp -r {{/path/to/source/folder/*}} {{/path/to/target/folder}}`
+
+- Copy of a file adding an extension:
```
This last two could be squashed into a single one.

waldyrious replied:
Agreed. Any suggestions on what to change/keep?

ostera replied:
I like the idea of using shell expansion, as I use them all the time (`!!:s/this/that/` is a life savior), but they are a bit specific to whatever the user is running.

I'd rather have `cp file.html file.html.backup` which is slightly more readable and covers both cases (since the backup could clearly be named anything), without digging into what `{{,.backup}}` actually implies.

waldyrious replied:
In that case we could simply remove the last two examples, since the first example covers this. I'd rather keep one of the -r examples as they're related to cp's functionality and may not be as easily deduced.

waldyrious commented on pages/common/cp.md:
```diff
@@ -1,27 +1,35 @@
+- Copy a folder recursively to another location:
 
-`cp -r {{/path/to/original}} {{/path/to/copy}}`
+`cp {{/path/to/folder}} {{/path/to/copy}}`
```
Ha! Good catch :) I'll change.

agnivade commented:
Since we are on this page, can we add the `-i` option too ? Seems like a useful flag to add.

waldyrious commented:
We could add -i, sure, but we first need to resolve the existing space issues 😄

waldyrious commented:
Ok guys, I've removed the last 2 examples which were shell features rather than cp-specific features, and added the missing `-r` as @ostera suggested. Since that means we won't have the excess space issues, I think it could be merged as-is, i.e. including two examples for file copy and their duals for directory copy.

agnivade commented:
Can we add `-i` now ? :stuck_out_tongue_closed_eyes:

waldyrious commented:
@agnivade sure -- any suggestions for a description?

agnivade commented:
Something on the lines of "Interactively copy files or folders from target to destination"

waldyrious commented:
Hmm, maybe it would be nice to use a glob pattern for this one for a change, what do you think?

agnivade commented:
Sure.

waldyrious commented:
@agnivade PTAL :)

waldyrious commented:
Hurray! After almost an year, this PR will get merged :)

ps - I rebased the commits interactively in the command line, to clean up the branch a bit before merging.
